### ISSUE TITLE
headless Xvfb or xvfb-run for electron how-to config into custom or standard build works on start script? 

### ISSUE THREAD

**Issue by kmacpher67** [cloudfoundry/nodejs-buildpack#89]
What version of Cloud Foundry and CF CLI are you using? (i.e. What is the output of running `cf curl /v2/info && cf version`?

{
   "name": "",
   "build": "",
   "support": "http://support.cloudfoundry.com",
   "version": 0,
   "description": "Cloud Foundry",
   "authorization_endpoint": "https://login.cloud.pcftest.com",
   "token_endpoint": "https://uaa.cloud.pcftest.com",
   "min_cli_version": null,
   "min_recommended_cli_version": null,
   "api_version": "2.80.0",
   "app_ssh_endpoint": "ssh.cloud.pcftest.com:2222",
   "app_ssh_host_key_fingerprint": "3e:d9:f9:02:29:9d:f6:4b:f2:90:fe:4b:05:85:35:8d",
   "app_ssh_oauth_client": "ssh-proxy",
   "logging_endpoint": "wss://loggregator.cloud.pcftest.com:4443",
   "doppler_logging_endpoint": "wss://doppler.cloud.pcftest.com:4443"
}

>cf --version
cf version 6.22.2+a95e24c-2016-10-27

What version of the buildpack you are using?

buildpack: https://github.com/cloudfoundry/nodejs-buildpack#v1.5.21

    "node": "~4.6.0",
    "npm": "2.x"

If you were attempting to accomplish a task, what was it you were attempting to do?

run electron headless for print PDF 

What did you expect to happen?

have it run in CF instance like it doesn't on my linux Ubuntu Desktop version. 

What was the actual behavior?

Nada happens. 
Tried building a custom forked version: 
https://github.com/kmacpher67/nodejs-buildpack.git#xvfb

Got the vxfbrun to work from compile, not able to have it run from instance on cf: 
`dist$ cf start cdchimay-ClientReportsServer-Electron`

restage / start deploy the app and the 
`-----> Downloaded app buildpack cache (94M)` 
in the compile script the  `xvfb-run` is found and run. but it never shows up in the vcap filesystem. 

See the branch above for the details on configuration.  

Please confirm where necessary:
* [ ] I have included a log output
* [ ] My log includes an error message
* [ x] I have included steps for reproduction
the following command fails to find `xvfb-run`
`    "start": "xvfb-run --server-args='-screen 0, 1024x768x16' node server/server.js",`

When it tries to start I get a big red "xvfb-run" not found.

**Comment by dgodd** [comment]
@kmacpher67 Can I clarify the problem you are having? It seems to me that electron is unrelated to your request, and that your actual request is that xvfb be installed? 

If so, a possibility for you is to vendor xvfb in your app

**Comment by kmacpher67** [comment]
Right, it could be node or java cf buildpack, but code that I'm running needs a visual frame buffer to paint against.  In this case as you explain it could be node electron, but in other cases I've had j2ee apps that needed a xvfb to use the graphics libs. 
essentially the answer if I owned sudo would be 
`sudo apt-get install xvfb libgtk2.0-0 libgconf-2-4 libasound2 libxtst6 libxss1` 
and i'm likely going to need a few more goodies (like fonts) in there too. 

So in luie of a 5 seconds worth of `sudo apt-get` I'm going down the painful highway of trying to figure out the voodoo that is buildpack lifecycles on cloudfoundy. 
Here's the compile from the standard build that I modified: 
https://github.com/kmacpher67/nodejs-buildpack/blob/xvfb/bin/compile
basically I d/l the deb file: 
`BUILDPACK_XVFB_PACKAGE=http://archive.ubuntu.com/ubuntu/pool/main/x/xorg-server/xvfb_1.15.1-0ubuntu2.7_amd64.deb`
copy it into a directory that I should be able to see from the app after the droplet deploys: 
`/home/vcap/app/bin` but alas `xvfb-run` executable is not there in that directory. 

So how do I install xvfb into cloudfoundry app using the buildpack stuff.

**Comment by kmacpher67** [comment]
Figured it was that the xvfb-run wasn't in the path, so we tried hard coding it: 
```2017-05-25T17:23:01.27-0400 [App/0] OUT > pr.reportsserver@1.5.23 start /home/vcap/app
2017-05-25T17:23:01.27-0400 [App/0] OUT > /home/vcap/app/bin/xvfb-run node server/server.js
2017-05-25T17:23:01.27-0400 [App/0] ERR sh: 1: /home/vcap/app/bin/xvfb-run: not found
2017-05-25T17:23:01.29-0400 [App/0] ERR npm ERR! argv "/home/vcap/app/.heroku/node/bin/node" "/home/vcap/app/.heroku/node/bin/npm" "start"
2017-05-25T17:23:01.29-0400 [App/0] ERR npm ERR! Failed at the xx.printpdf@1.5.23 start script '/home/vcap/app/bin/xvfb-run node server/server.js'.
2017-05-25T17:23:01.29-0400 [App/0] ERR npm ERR!     npm bugs xx.printpdf
2017-05-25T17:23:01.29-0400 [App/0] ERR npm ERR! Linux 4.4.0-72-generic```

**Comment by kmacpher67** [comment]
https://github.com/kmacpher67/nodejs-buildpack/tree/xvfb2
Created a new branch based on the heroku apt-get style build from standard package builder. 
Somebody from Pivotal did a pivotal CF re-mix: 
https://github.com/pivotal-topher-bullock/heroku-xvfb-buildpack/tree/master/bin
Utilized that to some success to get apt-get style build/ path, ld_library_path etc. working:
In our project we have `Aptfile` having these values: 
```
http://archive.ubuntu.com/ubuntu/pool/main/x/xorg-server/xvfb_1.15.1-0ubuntu2.7_amd64.deb
http://archive.ubuntu.com/ubuntu/pool/main/x/xauth/xauth_1.0.7-1ubuntu1_amd64.deb
x11-xkb-utils
xfonts-base
xfonts-75dpi
xfonts-scalable
libxfont1
xserver-common
libaudit1
libc6
libgcrypt11
libgl1-mesa-glx
libpixman-1-0
libselinux1
libxau6
libxdmcp6
libgl1-mesa-dri
libxpm4
libxrender1
libgtk2.0-0
libnss3
libgconf-2-4
x11-apps
libxss1
libnss3-dev
gcc-multilib
g++-multilib
libgtk2.0-dev
```
However the "scripts": {
`"start": "XAUTHORITY=/tmp/xvfb-run.2NG0xl/Xauthority xvfb-run --server-args='-screen 0, 1024x768x16' node server/server.js",  ` 
doesn't work get error: 
```
start /home/vcap/app
2017-05-30T10:42:41.58-0400 [App/0]      OUT > XAUTHORITY=/tmp/xvfb-run.2NG0xl/Xauthority xvfb-run --server-args='-screen 0, 1024x768x16' node server/server.js
2017-05-30T10:42:41.60-0400 [App/0]      ERR     xvfb-run: error: Xvfb failed to start
2017-05-30T10:42:41.62-0400 [App/0]      ERR npm ERR! Linux 4.4.0-72-generic
2017-05-30T10:42:41.62-0400 [App/0]      ERR npm ERR! argv "/home/vcap/app/.heroku/node/bin/node" "/home/vcap/app/.heroku/node/bin/npm" "start"
2017-05-30T10:42:41.62-0400 [App/0]      ERR npm ERR! node v4.6.0
2017-05-30T10:42:41.62-0400 [App/0]      ERR npm ERR! npm  v2.15.12
2017-05-30T10:42:41.62-0400 [App/0]      ERR npm ERR! code ELIFECYCLE
2017-05-30T10:42:41.62-0400 [App/0]      ERR npm ERR! pr.reportsserver@1.5.23 start: `XAUTHORITY=/tmp/xvfb-run.2NG0xl/Xauthority xvfb-run --server-args='-screen 0, 1024x768x16' node server/server.js` ....
2017-05-30T10:42:41.63-0400 [App/0]      ERR npm ERR!     /home/vcap/app/npm-debug.log
```

How do I get the debug for this? 
`/home/vcap/app/npm-debug.log` 
or for the Xvfb service?

**Comment by dgodd** [comment]
I suspect that the problem is not actually npm, but is xvfb-run . Thus npm-debug.log will not help. 

I would suggest that debugging your buildpack is probably easiest by using docker to run the cloudfoundry/cfllinuxfs2 docker image, and mounting your app and buildpack inside the docker image, and running `/bin/compile` manually, and then running the above start command manually.

**Comment by kmacpher67** [comment]
it was a path issue on the install environment. 
had to add "/app" to before all the ".apt" directories for it to work. (plus had to add the node npm install too) 
Then I found this error: 
```
vcap@79d27834-9e34-4f4a-4d02-c16f7b5e79f4:~$ xvfb-run  -e /dev/stdout node server/server.js
Xvfb: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory
xvfb-run: error: Xvfb failed to start
```
LIB DEPENDENCY HELL LEVEL 7. 
```
/home/vcap
vcap@395f1827-e2e9-4bfe-4021-28835e7a6116:~$ find /home/vcap/app/.apt/usr -name "libGL*"
/home/vcap/app/.apt/usr/lib/x86_64-linux-gnu/libGLU.so.1.3.1
/home/vcap/app/.apt/usr/lib/x86_64-linux-gnu/libGL.so
/home/vcap/app/.apt/usr/lib/x86_64-linux-gnu/libGLU.so
/home/vcap/app/.apt/usr/lib/x86_64-linux-gnu/libGLU.a
/home/vcap/app/.apt/usr/lib/x86_64-linux-gnu/mesa/libGL.so
/home/vcap/app/.apt/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1.2.0
/home/vcap/app/.apt/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
/home/vcap/app/.apt/usr/lib/x86_64-linux-gnu/libGLU.so.1
vcap@395f1827-e2e9-4bfe-4021-28835e7a6116:~$```

**Comment by kmacpher67** [comment]
`xvfb-run  -e /dev/stdout node server/server.js`
The -e help show details. 
``` 
Initializing built-in extension GLX
[dix] Could not init font path element /usr/share/fonts/X11/misc, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/cyrillic, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/100dpi/:unscaled, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/75dpi/:unscaled, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/Type1, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/100dpi, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/75dpi, removing from list!
sh: 1: /usr/bin/xkbcomp: not found
sh: 1: /usr/bin/xkbcomp: not found
XKB: Failed to compile keymap
Keyboard initialization failed. This could be a missing or incorrect setup of xkeyboard-config.
(EE)
Fatal server error:
(EE) Failed to activate core devices.(EE)
xvfb-run: error: Xvfb failed to start
vcap@395f1827-e2e9-4bfe-4021-28835e7a6116:~$
```
However, the keyboard file is there in the custom install directory structure: 
`xkbcomp`
```vcap@395f1827-e2e9-4bfe-4021-28835e7a6116:~/app/.apt/usr/bin$ ls -ltr xkbcomp
-rwxr-xr-x 1 vcap vcap 201784 Jan  6  2014 xkbcomp
vcap@395f1827-e2e9-4bfe-4021-28835e7a6116:~/app/.apt/usr/bin$```
It brings up a key issue with the custom install directory structure for X.11 and xvfb for example look at the `/usr/share/fonts/X11/ misc... `not being found.  but they are there in the custom install directories: 
```vcap@395f1827-e2e9-4bfe-4021-28835e7a6116:~/app/.apt/usr$ ls share/fonts/X11/
75dpi  cyrillic  encodings  misc  Type1  util
``` 
PROBLEM: No longer dependency hell. It's CF - PAAS - container land sudo restrictions. 
How doesth one configure xvfb X.11 etc for a custom install directory structure?  
This is inherently a container only issue.  I'd like to say I've never done this previously, but I vaguely remember something similar to this with HPUX or AIX at a clients site where the enterprise sys admins locked us out or maybe they didn't those are crappy *nix implementations. 
Any suggestions, my google ninja skillz are at a loss for the correct fuzzy search key words. 
Apparently, there is no formal taxonomy search lexicons for archaic *nix X.11 path-ing issues.

**Comment by sclevine** [comment]
Might be worth looking here: https://github.com/pivotal-cf-experimental/apt-buildpack/blob/master/bin/compile

**Comment by kmacpher67** [comment]
Currently got it this far: 
https://gist.github.com/fxtentacle/960cdb96ece01add8686
added this and it allowed the problem with the 
`sh: 1: /usr/bin/xkbcomp: not found`
Now it looks like the Xvfb is running on :99 but the PDF generation is still not working. 
So now I'm trying to figure out settings. 
```
vcap@33ae328c-c9a9-46ce-4d1f-b822b270d205:~$ ps -ef |grep -i xvfb
vcap          13       0  0 18:44 ?        00:00:00 bash -c pwd; ls -ltra; . ~/.profile.d/000_apt.sh; xvfb-run --auto-servernum -e /dev/stdout npm run startSimple
vcap          39      13  0 18:44 ?        00:00:00 /bin/sh /home/vcap/app/.apt/usr/bin/xvfb-run --auto-servernum -e /dev/stdout npm run startSimple
vcap          51      39  0 18:44 ?        00:00:00 Xvfb :99 -screen 0 640x480x8 -nolisten tcp -auth /home/vcap/tmp/xvfb-run.RT2QDl/Xauthority
vcap        2495    2409  0 20:46 pts/0    00:00:00 grep --color=auto -i xvfb
vcap@33ae328c-c9a9-46ce-4d1f-b822b270d205:~$
``` 
So now I'm trying to get the settings to work, still can get the PATH to work correctly inside npm/node package.json, but I call it before in the buildpack `release` seems to be started as testimony by the output above. 

Finally there were a handful of packages missing to get electron to run
Add this to the apt-get file `Aptfile` note: that file goes in your project deployment directory. 
```
x11-common
libx11-6
libasound2
http://ubuntu.mirrors.tds.net/ubuntu/pool/main/libx/libxtst/libxtst6_1.2.2-1_amd64.deb
```

**Comment by kmacpher67** [comment]
The one fix has a custom profile script, I moved that up above the calls for the heroku node/npm installs. 
I wonder why not use the apt-get to install these and on consistent method for creating binaries. 

note there are two different .profile.d scripts:  
`Writing profile script .profile.d/000_apt.sh`
`Writing profile script .profile.d/001_apt.sh`  <--- as the `/app` directory added to paths for ssh on CF. 

to get stuff working testing from a command line $ after using cf ssh <instance> 
$ ` . ./app/.profile.d/001_apt.sh` 
Still get a message about the RANDR stuff. 
```vcap@e900f0a6-1006-47aa-41da-c7e8:~$ electron --version
Xlib:  extension "RANDR" missing on display ":99".
v1.6.10
vcap@e900f0a6-1006-47aa-41da-c7e8:~$``` 
Various other complaints by X during the rendering. but for the most part xwindows stuff works. 
The `Xvfb` is working and configured correctly if you can do a `xclock` and it just sits there with no errs. The clock is being rendered on the "virtual" buffer so you don't see it, but the fact you're able to run the graphics program without errors blowing up means you got this portion of the game configured. 

There was likely other crap too that I figured out on the fly.  This is a Frankenstein merge of multiple snippets <Literally> from various sites and url references. 
1> click fork on the build-pack your looking to mutilate for X.11 
2> make a branch for xvfb 
2> copy these stuff into your various goodies https://github.com/pivotal-topher-bullock/heroku-xvfb-buildpack
3> ref this pivotal guys experimental stuff for CF specific paths crap??: 
4> use this guys sym-links to fix X.11 hard coded paths! https://gist.github.com/alonisser/11192482
5> use this `Aptfile` list in your core project make sure it's in the main top deployment directory. 
```xvfb
xauth
x11-xkb-utils
xfonts-base
xfonts-75dpi
xfonts-100dpi
xfonts-scalable
libxfont1
xserver-common
libaudit1
libc6
libgcrypt11
libglu1-mesa
libgl1-mesa-glx
libgl1-mesa-dri
libpixman-1-0
libselinux1
libxau6
libxdmcp6
libxpm4
libxrender1
libgtk2.0-0
libnss3
libgconf-2-4
x11-apps
libxss1
libnss3-dev
gcc-multilib
g++-multilib
libgtk2.0-dev
x11-xserver-utils
``` 
finally, I could not get the `Xvfb` or `xvfb-run` to work "inside" npm-node it seemed like the paths were all screwy.  So that's why I start the Xvfb server and specify the DISPLAY=:99 and make sure that's passed into the `DISPLAY=:99 npm start`
I'm not a linux guy apparently, I know enough to be dangerous. If you got any recommendations or details let me know. 
This would basically work for any graphics X.11 application running on CF where non-root X.11 is required.

**Comment by sclevine** [comment]
Closing due to inactivity.

**Comment by kmacpher67** [comment]
I built a custom build pack at the time.
It worked fine with Ubuntu 14 LTS until it broke horribly bad with upgrade to Ubuntu 16, we moved to docker images and somebody else figured out the dependency hell.
